src/librustc_target/spec/armv7r_none_eabihf.rs

```diff
@@ -0,0 +1,40 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Targets the Little-endian Cortex-R4F/R5F processor (ARMv7-R)
+
+use std::default::Default;
+use spec::{LinkerFlavor, PanicStrategy, Target, TargetOptions, TargetResult};
+
+pub fn target() -> TargetResult {
+    Ok(Target {
+        llvm_target: "armv7r-none-eabihf".to_string(),
+        target_endian: "little".to_string(),
+        target_pointer_width: "32".to_string(),
+        target_c_int_width: "32".to_string(),
+        data_layout: "e-m:e-p:32:32-i64:64-v128:64:128-a:0:32-n32-S64".to_string(),
+        arch: "arm".to_string(),
+        target_os: "none".to_string(),
+        target_env: "".to_string(),
+        target_vendor: "".to_string(),
+        linker_flavor: LinkerFlavor::Gcc,
+
+        options: TargetOptions {
+            executables: true,
+            relocation_model: "static".to_string(),
+            panic_strategy: PanicStrategy::Abort,
+            features: "+vfp3,+d16,+fp-only-sp".to_string(),
```

Comment:
The `armebv7r-none-eabihf` target has a `+v7` here. Why does that target need it and this one doesn't? I would expect LLVM to enable the `v7` feature if the `llvm-triple` includes the word `arm*v7` in it.

---

Semi on-topic: anyone knows how to make llc, rustc, or in general LLVM, output all the *enabled* target features for a *particular* triple? Something like `llc -triple=armebv7r-none-eabihf --print enabled-target-features` (which of course that doesn't work) that prints `+v7,+rclass,+dsp,..`. I know that `llc -mattr=help -march=arm` prints all the target features that makes sense for the ARM architecture but I want to know which ones are enabled. `rustc --print cfg` *almost* does what I want but doesn't include target features that have not been whitelisted. cc @parched @nagisa

Reply:
During the development process, I pulled the list of features from [here](https://github.com/llvm-mirror/llvm/blob/master/lib/Target/ARM/ARM.td) when I thought the fault that I was encountering was unrelated to the issue that https://github.com/alexcrichton/cc-rs/pull/339 addresses. That being said, (L422-425)[https://github.com/llvm-mirror/llvm/blob/master/lib/Target/ARM/ARM.td#L422-L425] of that file actually define the [sub]features that enabling that feature enables. I was a little verbose (as mentioned) and chose to be very explicit about enabling the feature groups AND individual features, because I wasn't entirely sure where Rust was pulling its version of LLVM from at that time. There are, however, specific features for the [ARMv7r](https://github.com/llvm-mirror/llvm/blob/master/lib/Target/ARM/ARM.td#L584-L588) processors, as well as for the [Cortex R5, specifically](https://github.com/llvm-mirror/llvm/blob/master/lib/Target/ARM/ARM.td#L906-L913).